cellxgene_schema_cli/cellxgene_schema/atac_seq.py

```diff
@@ -0,0 +1,370 @@
+    elif not ((df == "u").all() or (df == "p").all()):
+        return "Anndata.obs.assay_ontology_term_id has mixed paired and unpaired terms."
+
+
+def validate_anndata_is_primary_data(anndata_file: str) -> Optional[str]:
+    with h5py.File(anndata_file) as f:
+        is_primary_data = ad.io.read_elem(f["obs"])["is_primary_data"]
+    if not is_primary_data.all():
+        return "Anndata.obs.is_primary_data must all be True."
+
+
+def validate_anndata_feature_reference(anndata_file: str) -> Optional[str]:
+    with h5py.File(anndata_file) as f:
+        feature_reference = ad.io.read_elem(f["var"])["feature_reference"]
+        unique_organism_ontology_term_ids = ad.io.read_elem(f["obs"])["organism_ontology_term_id"].unique()
+    # check that var["feature_reference"] is equal to "NCBITaxon:9606" xor "NCBITaxon:10090"
+    if not (feature_reference == "NCBITaxon:9606").all() ^ (feature_reference == "NCBITaxon:10090").all():
+        return "Anndata.var.feature_reference must be either NCBITaxon:9606 or NCBITaxon:10090."
+    # check that var["feature_reference"] and obs["organism_ontology_term_id"] have the same unique values
+    unique_feature_reference = feature_reference.unique()
+    if unique_organism_ontology_term_ids != unique_feature_reference:
+        return "Unique Anndata.obs.organism_ontology_term_id must be equal to unqiue Anndata.var.feature_reference."
+
+
+def detect_chromosomes(parquet_file: str) -> list[str]:
+    logger.info("detecting chromosomes")
+    df = ddf.read_parquet(parquet_file, columns=["chromosome"]).drop_duplicates()
+    return df["chromosome"].values.compute()
+
+
+def index_fragment(fragment_file: str, parquet_file: str, tempdir: str, override_write_algorithm: Optional[str] = None):
+    # sort the fragment by chromosome, start coordinate, and stop coordinate, then compress it with bgzip
+    bgzip_output_file = fragment_file.replace(".gz", ".bgz")
+    bgzip_output_path = Path(bgzip_output_file)
+    bgzip_output_path.unlink(missing_ok=True)
+    bgzip_output_path.touch()
+    bgzip_write_lock = dd.Lock()  # lock to preserver write order
+
+    if override_write_algorithm:
+        write_algorithm = write_algorithm_by_callable[override_write_algorithm]
+    elif not shutil.which("bgzip"):  # check if bgzip cli is installed
+        logger.warning("bgzip is not installed, using slower pysam implementation")
+        write_algorithm = write_algorithm_by_callable["pysam"]
+    else:
+        write_algorithm = write_algorithm_by_callable["cli"]
+
+    chromosomes = detect_chromosomes(parquet_file)
+    jobs = prepare_fragment(chromosomes, parquet_file, bgzip_output_file, tempdir, bgzip_write_lock, write_algorithm)
+    # limit calls to dask.compute to improve performance. The number of jobs to run at once is determined by the
+    # step variable. If we run all the jobs in the same call to dask.compute, the local cluster hangs.
+    # TODO: investigate why
+    step = 4
```

Comment:
how was this number determined? are we planning on running this many cores for the atac-seq validation container?

Reply:
This is number of jobs we send to the dask scheduler at a time. The dask configuration will determine the compute used. The value of 4 was determined by running the sample fragment a few times with a decreasing number of jobs until it didn't freeze.